### PR TITLE
feat(components): improve hover states for buttons

### LINKED ISSUE #19965: bug: Menu button is not using proper color due to :host-context on iOS Safari
# Bug Report

**Ionic version:**

[x] **4.x**

**Current behavior:**

View on iOS:

https://sean-preconceptiontest.tinytake.com/media/b68647?filename=1574354453447_iOS+Screenshot.png&sub_type=thumbnail_preview&type=attachment&width=370&height=798

Elements on iOS:

https://sean-preconceptiontest.tinytake.com/media/b68629?filename=1574354107689_iOS+Elements.png&sub_type=thumbnail_preview&type=attachment&width=1200&height=868

**Expected behavior:**

Color should changed based on background color on iOS

View: (on chrome) same as android 10 on Essential Phone

https://sean-preconceptiontest.tinytake.com/media/b68643?filename=1574354314403_desired+View.png&sub_type=thumbnail_preview&type=attachment&width=356&height=799

Elements on Chrome:

https://sean-preconceptiontest.tinytake.com/media/b6862d?filename=1574354172989_chrome+elements.png&sub_type=thumbnail_preview&type=attachment&width=1199&height=683

**Steps to reproduce:*

```
export function WithMenu(props: IProps) {
  return (
    <IonHeader>
      <IonToolbar>
        <IonMenuButton slot="start" />
        <IonTitle>{props.pageTitle}</IonTitle>
      </IonToolbar>
    </IonHeader>
  );
}
```

 ```--ion-toolbar-background: #003366;
  --ion-toolbar-border-color: #ffffff;
  --ion-toolbar-color: #ffffff;
  --ion-toolbar-color-activated: #ccb7fc;
  --ion-toolbar-color-unchecked: #165499;
  --ion-toolbar-color-checked: #1a8aac;

```

**Ionic info:**

```
Ionic:

   Ionic CLI       : 5.4.6 (C:\Users\sagui\AppData\Roaming\npm\node_modules\ionic)
   Ionic Framework : @ionic/react 4.11.5

Utility:

   cordova-res : not installed
   native-run  : not installed

System:

   NodeJS : v12.13.0 (C:\Program Files\nodejs\node.exe)
   npm    : 6.12.0
   OS     : Windows 10
```

### LINKED ISSUE #20213: Changing ionic menu color for ios
[x] 4.11.5

[x] bug report
[ ] feature request

**Current behavior:**
I run the ionic start command for the side menu starter, and add these lines to the variables.scss file, inside of the :root object:
```
  --ion-toolbar-background: var(--ion-color-primary);
  --ion-toolbar-color: var(--ion-color-light);
```
For android, the ion-menu-button and ion-back-button change to the toolbar color property, but on ios they are still set to primary, making the buttons invisible.

I am arguing this is a bug because of the varied behaviors between ios and android

**Expected behavior:**
I would expect the colors to be the same between ios and android

**Steps to reproduce:**
ionic starter: side menu
Add this to the variables.scss file under :root object:
  --ion-toolbar-background: var(--ion-color-primary);
  --ion-toolbar-color: var(--ion-color-light);
Run in ios and android, they will be different

**Related code:**

**Other information:**

**Ionic info:** 
```
Ionic:

   Ionic CLI                     : 5.4.13 (/usr/local/lib/node_modules/ionic)
   Ionic Framework               : @ionic/angular 4.11.8
   @angular-devkit/build-angular : 0.801.3
   @angular-devkit/schematics    : 8.1.3
   @angular/cli                  : 8.1.3
   @ionic/angular-toolkit        : 2.1.2

Cordova:

   Cordova CLI       : 9.0.0 (cordova-lib@9.0.1)
   Cordova Platforms : not available
   Cordova Plugins   : not available

Utility:

   cordova-res : 0.8.1
   native-run  : 0.3.0

System:

   Android SDK Tools : 26.1.1 (/Users/lucas.crandle/Library/Android/sdk)
   ios-deploy        : 2.0.0
   ios-sim           : 8.0.2
   NodeJS            : v12.14.1 (/usr/local/bin/node)
   npm               : 6.13.4
   OS                : macOS Mojave
   Xcode             : Xcode 11.3 Build version 11C29```

### PR DESCRIPTION
**Pull request by brandyscarney**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [x] Feature
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

Currently users have to know the exact opacity that the MD spec uses in order to change the hover color. This allows them to change the background without having to know the opacity. There is also a bug with dark mode because the hover state would need to be set to the toolbar color with the opacity for each fill. 

References #18279 Fixes #20213 Fixes #19965

## What is the new behavior?

- greatly reduces the requirement by users to set the background hover for dark modes, also eliminates the need to know what the hover opacity is for each button fill
- updates the MD dark theme per their spec

### Component Updates

Any component in the below table needs to have hover corrected.

Table Key:

- **Existing Hover (`master`)**: Did the component have the variables for `--background-hover` on `master`
- **Updated Hover?**: Have the hover variables been updated for this branch 

:white_large_square: - Incomplete
:white_check_mark: - Done
:cyclone: - Doesn't exist

| Name                   | Existing Hover (`master`) | Updated Hover?           | Updated Activated?       | Updated Focused?         |
| -----------------------| ------------------------- | ------------------------ | ------------------------ | ------------------------ |
| Action Sheet (Buttons) | :white_large_square:      | :white_check_mark:       | :white_check_mark:       | :white_check_mark:       |
| Back Button            | :white_check_mark:        | :white_check_mark:       | :cyclone:                | :white_check_mark:       |
| Button                 | :white_check_mark:        | :white_check_mark:       | :white_check_mark:       | :white_check_mark:       |
| FAB Button             | :white_check_mark:        | :white_check_mark:       | :white_check_mark:       | :white_check_mark:       |
| Item                   | :white_check_mark:        | :white_check_mark:       | :white_check_mark:       | :white_check_mark:       |
| Menu Button            | :white_check_mark:        | :white_check_mark:       | :cyclone:                | :white_check_mark:       |
| Segment Button         | :white_check_mark:        | :white_check_mark:       | :cyclone:                | :white_check_mark:       |
| Tab Button             | :white_large_square:      | :white_large_square:     | :white_large_square:     | :white_check_mark:     |

Remaining components here: https://github.com/ionic-team/ionic/issues/18279

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

Will go into Ionic 5. 

#### Activated Class

The `activated` class that is automatically added to buttons on press has been renamed to `ion-activated`. This will be more consistent with our `ion-focused` class we add and also will reduce conflicts with user's CSS.

#### CSS Variables

The `--background-hover` CSS variable on buttons will now have an opacity automatically set. Anyone setting it like the following:

```
--background-hover: rgba(44, 44, 44, 0.08);
```

Will likely not see a hover state anymore, it should be updated to the following:

```
--background-hover: rgba(44, 44, 44);
```

If the opacity desired is something other than what is set by default, use:

```
--background-hover: rgba(44, 44, 44);
--background-hover-opacity: 1;
```

### CODE REVIEW

**Comment by brandyscarney** [comment]
@liamdebeasi 

> I am going through all of these now, but wanted to provide some initial feedback on action sheet so you're not waiting on me to finish my review:
> 
> Action Sheet
> 
> 1. iOS only: `--button-background` does not seem to work. `background-color: transparent` appears to be overriding it.

Fixed by: https://github.com/ionic-team/ionic/pull/19440/commits/1d4fe52af6671b4874d2dd73d7d220f78aa7ea0d

File: BREAKING.md [hunk 1]
```diff
@@ -98,9 +104,101 @@ See the [CSS Utilities responsive display documentation](https://ionicframework.
...
+
+```css
+--background-activated-opacity
+--background-focused-opacity
+--background-hover-opacity
+```
+
+It also updates the Action Sheet component so that the variables will be prefixed with `button`. See the [Action Sheet](#action-sheet) section in this document for all of the variable names.
+
+This allows you to still have the control over the opacity if desired, but when updating the state, you only have to set the main variables: `--background-activated`, `--background-focused`, `--background-hover` and the button will still match the spec. This is most important when changing the global theme, as updating the toolbar color will automatically update the hover states for all of the buttons in a toolbar, regardless of their fill & without having to know what each opacity is.
```
Comment (liamdebeasi) [inline BREAKING.md]:
```suggestion
This allows you to still have control over the opacity if desired, but when updating the state, you only have to set the main variables: `--background-activated`, `--background-focused`, `--background-hover` and the button will still match the spec. This is most important when changing the global theme, as updating the toolbar color will automatically update the hover states for all of the buttons in a toolbar, regardless of their fill and without having to know what each opacity is.
```

File: BREAKING.md [hunk 2]
```diff
@@ -98,9 +104,101 @@ See the [CSS Utilities responsive display documentation](https://ionicframework.
...
+  --button-background-hover-opacity: 0.5;
+}
+
+/* Setting the fab button background to use the text color with the default opacity on md */
+.md ion-fab-button {
+  --color: #222;
+  --background-hover: #222;
+}
+
+/* Setting the theme background for items to use #000 (black) with the default theme opacity */
```
Comment (liamdebeasi) [inline BREAKING.md]: Why not just use `black` instead of doing `#000` and explaining that it is black?

File: BREAKING.md [hunk 3]
```diff
@@ -254,7 +397,78 @@ The `inputmode` property for `ion-searchbar` now defaults to `undefined`. To get
 
 #### Segment
 
-<!-- TODO https://gist.github.com/brandyscarney/e6cfe43c359bb2c932e12f8d615e1669 -->
+Segment was completely revamped to use the new iOS design including an all new gesture that applies for both Material Design & iOS. Due to these changes, some breaking changes were inevitably introduced in order to support the new design.
```
Comment (liamdebeasi) [inline BREAKING.md]:
```suggestion
Segment was completely revamped to use the new iOS design including an all new gesture that applies for both Material Design and iOS. Due to these changes, some breaking changes were inevitably introduced in order to support the new design.
```

File: BREAKING.md [hunk 4]
```diff
@@ -254,7 +397,78 @@ The `inputmode` property for `ion-searchbar` now defaults to `undefined`. To get
...
+A `--background` variable has been added to style the `ion-segment` component. As a result of this, the following background variables for a child segment button must now be set on the `ion-segment-button`:
+
+```
+--background: Background of the segment button
+--background-checked: Background of the checked segment button
+--background-disabled: Background of the disabled segment button
+--background-hover: Background of the segment button on hover
+```
+
+> Note: iOS no longer checks the button background, so setting the `--background-checked` variable may have an undesired outcome. It uses an indicator to slide between the buttons. See the previous section on the indicator color variables.
```
Comment (liamdebeasi) [inline BREAKING.md]: What does "it" refer to here?

File: BREAKING.md [hunk 5]
```diff
@@ -254,7 +397,78 @@ The `inputmode` property for `ion-searchbar` now defaults to `undefined`. To get
 
 #### Segment
 
-<!-- TODO https://gist.github.com/brandyscarney/e6cfe43c359bb2c932e12f8d615e1669 -->
+Segment was completely revamped to use the new iOS design including an all new gesture that applies for both Material Design & iOS. Due to these changes, some breaking changes were inevitably introduced in order to support the new design.
+
+##### Button States
+
+- The activated styles and properties have been removed. These are no longer being used in the latest spec as the indicator and ripple are used to show activation. Properties removed:
```
Comment (liamdebeasi) [inline BREAKING.md]:
```suggestion
- The activated styles and variables have been removed. These are no longer being used in the latest spec as the indicator and ripple are used to show activation. Variables removed:
```

File: BREAKING.md [hunk 6]
```diff
@@ -329,6 +543,10 @@ The `width` property has been removed in favor of using CSS styling.
   </ion-split-pane>
   ```
 
+#### Tabs
+
+- [Focused State](#activated-focused-hover-states) have been updated.
```
Comment (liamdebeasi) [inline BREAKING.md]:
```suggestion
- [Focused States](#activated-focused-hover-states) have been updated.
```

File: BREAKING.md [hunk 7]
```diff
@@ -98,9 +104,101 @@ See the [CSS Utilities responsive display documentation](https://ionicframework.
 
 #### Activated, Focused, Hover States
 
-The `.activated` class that gets added has been renamed to `.ion-activated` for consistency with how we add focused to elements and to avoid conflicts in users' CSS.
+The `.activated` class that is automatically added to clickable components has been renamed to `.ion-activated` for consistency with how we add focused to elements and to avoid conflicts in users' CSS.
```
Comment (liamdebeasi) [inline BREAKING.md]:
```suggestion
The `.activated` class that is automatically added to clickable components has been renamed to `.ion-activated`.
```
I don't think an explanation adds much here

File: BREAKING.md [hunk 8]
```diff
@@ -254,7 +397,78 @@ The `inputmode` property for `ion-searchbar` now defaults to `undefined`. To get
...
+```
+--background: Background of the segment button
+--background-checked: Background of the checked segment button
+--background-disabled: Background of the disabled segment button
+--background-hover: Background of the segment button on hover
+```
+
+> Note: iOS no longer checks the button background, so setting the `--background-checked` variable may have an undesired outcome. It uses an indicator to slide between the buttons. See the previous section on the indicator color variables.
+
+The above variables *will not* be inherited in the button if set on the `ion-segment`. In addition to this, all color variables should also be set in the button for consistency:
```
Comment (liamdebeasi) [inline BREAKING.md]:
```suggestion
The above variables *will not* be inherited in the button if set on the `ion-segment`. In addition to this, all color variables should also be set on the button for consistency:
```

File: BREAKING.md (hunk 2 above)
Comment (brandyscarney) [reply BREAKING.md]: gp

Comment (brandyscarney) [reply BREAKING.md]: oh - because I think it makes more sense when you look at how it converts to rgba

File: BREAKING.md (hunk 5 above)
Comment (brandyscarney) [reply BREAKING.md]: I'll update this to "custom CSS properties" since that's what we call them in our docs

File: BREAKING.md (hunk 6 above)
Comment (brandyscarney) [reply BREAKING.md]: It's just one state for focused

Comment (liamdebeasi) [reply BREAKING.md]: `have` should probably say `has` in that case

**Comment by brandyscarney** [comment]
Ready 2 Merge

**Pull request merged**